contracts/debtAllocators/GenericDebtAllocator.sol

```diff
@@ -0,0 +1,211 @@
+        uint256 targetDebt = Math.min(
+            (_vault.totalAssets() * config.targetRatio) / MAX_BPS,
+            // Make sure it is not more than the max allowed.
+            params.max_debt
+        );
+
+        // If we need to add more.
+        if (targetDebt > params.current_debt) {
+            // We can't add more than the current idle.
+            uint256 toAdd = Math.min(
+                targetDebt - params.current_debt,
+                // This may underflow if idle is less than the min.
+                // Thats okay since it should return false anyway.
+                _vault.totalIdle() - _vault.minimum_total_idle()
+            );
+
+            // If the amount to add is over our threshold.
+            if (toAdd > config.minimumChange) {
+                // Return true and the calldata.
+                return (
+                    true,
+                    abi.encodeCall(
+                        _vault.update_debt,
+                        (_strategy, params.current_debt + toAdd)
+                    )
+                );
+            }
+            // If target debt is lower than the current.
+        } else if (targetDebt < params.current_debt) {
+            // Find out by how much.
+            uint256 toPull = Math.min(
+                params.current_debt - targetDebt,
+                // Account for the current liquidity constraints.
+                IVault(_strategy).maxWithdraw(address(_vault))
+            );
+
+            // Check if its over the threshold.
+            if (toPull > config.minimumChange) {
+                // If so return true and the calldata.
+                return (
+                    true,
+                    abi.encodeCall(
+                        _vault.update_debt,
+                        (_strategy, params.current_debt - toPull)
+                    )
+                );
+            }
+        } else {
+            return (false, bytes("No Change Needed"));
+        }
+    }
+
+    /**
+     * @notice Sets a new target debt ratio for a strategy.
+     * @dev A `minimumChange` for that strategy must be set first.
+     * This is to prevent debt to be updated to frequently.
+     *
+     * @param _strategy Address of the strategy to set.
+     * @param _targetRatio Amount in Basis points to allocate.
+     */
+    function setTargetDebtRatio(
+        address _strategy,
+        uint256 _targetRatio
+    ) external onlyGovernance {
+        // Make sure the strategy is added to the vault.
+        require(IVault(vault).strategies(_strategy).activation != 0, "!active");
+        // Make sure a minimumChange has been set.
+        require(configs[_strategy].minimumChange != 0, "!minimum");
+
+        // Get what will be the new total debt ratio.
+        uint256 newDebtRatio = debtRatio -
+            configs[_strategy].targetRatio +
+            _targetRatio;
+
+        // Make sure it is under 100% allocated
+        require(newDebtRatio <= MAX_BPS, "!max");
```

Comment:
nit: i always think of "!" as meaning "not". 
so "not max" is less accurate than something like "Ratio Too High".

Reply:
https://github.com/yearn/vault-periphery/pull/15/commits/21b8ff41c6b5848a617675584c28b81077eb389e